### PR TITLE
Lower the min thread number for reduction

### PR DESCRIPTION
**Pull request by nouiz**
@cheshire

### CODE REVIEW

File: xla/service/gpu/ir_emission_utils.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ inline constexpr int64_t WarpSize() { return 32; }
 
 // Need at least 1024 threads/block for reasonable tree reduction
```
Comment (cliffwoolley) [inline xla/service/gpu/ir_emission_utils.h:52]: The comment should also be updated to reflect the code change.

Comment (nouiz) [reply xla/service/gpu/ir_emission_utils.h]: Right. I made this PR quickly just to help Google test it internally.
I just updated it to be sure it isn't forgotten.

**Comment by cheshire** [comment]
Looking into benchmarks, so far seems to cause very minor perf regressions (1-2%)

**Comment by cheshire** [comment]
Our benchmark for T5X showed a sizeable regression with this change:

```
   0.74x,    ???  (clean: 15830.0 us,       ???;  dirty: 21339.0 us,       ???): hlo_mobilenet_v2_x1
   1.09x,    ???  (clean:  6228.0 us,       ???;  dirty:  5726.0 us,       ???): hlo_resnet200_x1
   1.25x,    ???  (clean: 22373.0 us,       ???;  dirty: 17846.0 us,       ???): hlo_stable_diffusion_x1
   0.86x,    ???  (clean:  3014.0 us,       ???;  dirty:  3517.0 us,       ???): hlo_t5x_no_scan_x1
```

**Comment by cliffwoolley** [comment]
@cheshire ..though a sizeable _progression_ for Stable Diffusion.  Apparently @nouiz is correct that we'll need to be a little fancier about detecting when to use the smaller vs larger block size.  Can we make this at least configurable somehow in the meanwhile?

**Comment by cheshire** [comment]
Yeah we can, probably it makes sense to at least move this selection to an HLO pass. We can also consider autotuning this (and then not using cases which spill registers?..)

**Comment by cliffwoolley** [comment]
@cheshire - sure; I'm just looking for some kind of stopgap before we do any major overhaul.

That said, in the end block size is only but one of many knobs; it's just an important one to avoid maxing out unnecessarily, as that forecloses other options.  For the cases of slowdown you saw, we could also grab Nsight Systems traces before/after and find the specific kernel(s) that slowed down, then look at Nsight Compute profiles of those specific kernels to find out _why_ they slowed down.  This may lead to other avenues for smoothing things out besides "don't let the block size be smaller".

**Comment by cheshire** [comment]
The slowdown is from a MobileNet, which is compiled from PyTorch torchbench collection: https://gist.github.com/cheshire/c80a325102bb7ccd7222880fa2c77526

**Comment by nouiz** [comment]
After offline discussion, we need to update this PR to have the number of threads depends on the ptxas version.
If it is 12.2 or higher, use 1024. Otherwise, use 512.

**Comment by nouiz** [comment]
Superseded by https://github.com/openxla/xla/pull/3432

**Pull request closed without merging**